src/main/objects/HarmonicFunction.js

```diff
@@ -1,34 +1,89 @@
+
+    this.functionName = params["functionName"];
+    this.degree = params["degree"] === undefined ? getDegreeFromFuntionName(this.functionName) : params["degree"];
+    this.position = params["position"];
+    this.revolution = params["revolution"] === undefined ? (params["down"] === true ? "1>" : "1") : params["revolution"];
+    this.delay = params["delay"] === undefined ? [] : params["delay"];
+    this.extra = params["extra"] === undefined ? [] : params["extra"];
+    this.omit = params["omit"] === undefined ? [] : params["omit"];
+    this.down = params["down"] === undefined ? false : params["down"];
+    this.system = params["system"];
+    this.mode = params["mode"] === undefined ? 'major' : params["mode"];
+
+    // mapping to ChordComponent
+    if(this.position !== undefined) this.position = cm.chordComponentFromString(this.position);
+    this.revolution = cm.chordComponentFromString(this.revolution);
+    for(var i=0; i<this.delay.length; i++){
+        this.delay[i][0] = cm.chordComponentFromString(this.delay[i][0]);
+        this.delay[i][1] = cm.chordComponentFromString(this.delay[i][1]);
     }
+    for(i=0; i<this.extra.length; i++) this.extra[i] = cm.chordComponentFromString(this.extra[i]);
+    for(i=0; i<this.omit.length; i++) this.omit[i] = cm.chordComponentFromString(this.omit[i]);
 
-    this.getBasicChordComponents = function () {
-        if(down) return ["1>", "3", "5>"];
-        return ["1", "3", "5"];
+
+    function getDegreeFromFuntionName(functionName){
+        return {"T":1, "S":4, "D":5}[functionName];
     }
 
-    this.getPossibleToDouble = function (){
+    this.getBasicChordComponents = function () {
+        var chordComponentManager = new ChordComponentManager.ChordComponentManager();
+        var scale = this.mode === 'major' ? new Scale.MajorScale("X") : new Scale.MinorScale("X");
+        var basicChordComponents;
+
+        if(!this.down) {
+            var thirdPitch = Utils.mod(scale.pitches[Utils.mod(this.degree + 1, 7)] - scale.pitches[Utils.mod(this.degree - 1, 7)], 12);
+            var fifthPitch = Utils.mod(scale.pitches[Utils.mod(this.degree + 3, 7)] - scale.pitches[Utils.mod(this.degree - 1, 7)], 12);
+
+
+            basicChordComponents = [chordComponentManager.chordComponentFromString("1"),
+                                    chordComponentManager.basicChordComponentFromPitch(thirdPitch),
+                                    chordComponentManager.basicChordComponentFromPitch(fifthPitch)];
+        }
+        else {
+            basicChordComponents = [chordComponentManager.chordComponentFromString("1>"),
+                chordComponentManager.chordComponentFromString("3>"),
+                chordComponentManager.chordComponentFromString("5")];
+        }
+
+        return basicChordComponents;
+    };
+
+    //additional rules
+    if(Utils.contains(this.extra, cm.chordComponentFromString("9")) && !Utils.contains(this.extra, cm.chordComponentFromString("7"))) this.extra.push("7");
```

Comment:
Tu warto sprawdzić, czy w teście z 1_HarmonicFunctionTests#delay9_8 dla pierwszego opóźnienia nie jest przypadkiem dodawane do extra 7. Można taką asercję dopisać do tego testu...

Reply:
To rozumiem że w tym wypadku nie powinno dorzucać 7 do extra?

Reply:
Zrobiłem jakiegoś hacka na to i chciałem się tylko upewnić że dalej działa po Twojej zmianie :)
Natomiast mam jeszcze jedną prośbę od razu: mógłbyś jeszcze dopisać taki kawałek kodu, który: 
Jeśli w extra mamy 5< lub 5> to dajemy z urzędu do omit 5? :)